api/v1_projects.inc

```diff
@@ -752,3 +753,116 @@ function _get_enabled_filter($query_params)
+}
+
+function api_v1_project_page(string $method, array $data, array $query_params)
+{
+    global $pguser;
+
+    try {
+        $proof_project = create_proof_project($data, $query_params);
+        $project_page = $data[":pagename"];
+        $page_state = $query_params['pagestate'] ?? null;
+
+        if (null === $page_state) {
+            throw new InvalidValue("No page state found in request.");
+        }
+        if (!in_array($page_state, Rounds::get_page_states())) {
+            throw new InvalidValue(sprintf("%s is not a valid page state", $page_state));
+        }
+        if ($page_state != $project_page->page_state) {
+            $err = sprintf(
+                _('Warning: Page  "%1$s" is no longer in state "%2$s"; it is now in state "%3$s".'),
+                $project_page->page_name,
+                $page_state,
+                $project_page->page_state
+            );
+            throw new ProjectPageInconsistentStateException($err);
+        }
+
+        $proof_project_page = new ProofProjectPage($proof_project, $project_page);
+        if ($method == "GET") {
+            return $proof_project_page->render_page_text();
+        }
+        // must be PUT
+        $page_action = $query_params['pageaction'] ?? null;
+        if (null === $page_action) {
+            throw new BadRequest("No page action found in request.");
+        }
+        switch ($page_action) {
+            case 'resume':
+                return $proof_project_page->pp_resume_page();
+            case 'checkin':
+                return $proof_project_page->attempt_checkin(receive_project_text_from_request_body('text'));
+            case 'save':
+                return $proof_project_page->save(receive_project_text_from_request_body('text'));
+            case 'revert':
+                return $proof_project_page->save_and_revert(receive_project_text_from_request_body('text'));
+            case 'abandon':
+                return $proof_project_page->returnToRound($pguser);
+            default:
+                throw new BadRequest("$page_action is not a valid page action.");
+        }
+    } catch (ProjectException $exception) {
+        throw new NotFoundError($exception->getMessage(), $exception->getCode());
+    } catch (UserAccessException $exception) {
+        throw new ForbiddenError($exception->getMessage(), $exception->getCode());
+    }
+}
+
+function create_proof_project(array $data, array $query_params)
+{
+    $project = $data[":projectid"];
+    $state = $query_params["state"] ?? null;
+    try {
+        if (null === $state) {
+            throw new InvalidValue("No project state found in request.");
+        }
+        if (!in_array($state, ProjectStates::get_states())) {
+            throw new InvalidValue("Invalid project state");
+        }
+        if ($state != $project->state) {
+            $err = sprintf(
+                _('Warning: Project "%1$s" is not in state "%2$s"; it is now in state "%3$s".'),
```

Comment:
Done this.
There are two similar strings in LPage.inc in get_indicated_LPage() and project_continuity_test(). Should we change them also?

Reply:
I think we should change the one in `project_continuity_test()` as it is project-state related (and let's remove the `\n` prefix before that string too). I'd leave the one in `get_indicated_LPage()` as-is since it's page-state related.